keel-bakery-plugin/src/main/kotlin/com/netflix/spinnaker/keel/bakery/artifact/BakeryLifecycleMonitor.kt

```diff
@@ -0,0 +1,120 @@
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventStatus
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventType
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventType.BAKE
+import com.netflix.spinnaker.keel.lifecycle.LifecycleMonitor
+import com.netflix.spinnaker.keel.lifecycle.LifecycleMonitorRepository
+import com.netflix.spinnaker.keel.lifecycle.MonitoredTask
+import com.netflix.spinnaker.keel.orca.ExecutionDetailResponse
+import com.netflix.spinnaker.keel.orca.OrcaExecutionStatus.BUFFERED
+import com.netflix.spinnaker.keel.orca.OrcaExecutionStatus.RUNNING
+import com.netflix.spinnaker.keel.orca.OrcaService
+import io.github.resilience4j.retry.Retry
+import io.github.resilience4j.retry.RetryConfig
+import io.vavr.control.Try
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.boot.context.properties.EnableConfigurationProperties
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+import java.time.Duration
+
+@Component
+@EnableConfigurationProperties(LifecycleConfig::class)
+class BakeryLifecycleMonitor(
+  override val monitorRepository: LifecycleMonitorRepository,
+  override val publisher: ApplicationEventPublisher,
+  override val lifecycleConfig: LifecycleConfig,
+  private val orcaService: OrcaService,
+  @Value("\${spinnaker.baseUrl}") private val spinnakerBaseUrl: String
+): LifecycleMonitor(monitorRepository, publisher, lifecycleConfig) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  override fun handles(event: LifecycleEvent): Boolean =
+    event.type == BAKE
+
+  override fun typeHandled(): LifecycleEventType = BAKE
+
+  override suspend fun monitor(task: MonitoredTask) {
+    kotlin.runCatching {
+       orcaService.getOrchestrationExecution(task.link, DEFAULT_SERVICE_ACCOUNT)
+    }
+      .onSuccess { execution ->
+        when {
+          execution.status == BUFFERED -> publishCorrectLink(task)
+          execution.status == RUNNING -> publishRunningEvent(task)
+          execution.status.isSuccess() -> publishSucceededEvent(task)
+          execution.status.isFailure() -> publishFailedEvent(task)
+          else -> publishUnknownEvent(task, execution)
+        }
+
+        if (execution.status.isComplete()) {
+          endMonitoringOf(task)
+        }
+      }
+      .onFailure { exception ->
+        log.error("Error fetching status for $task: ", exception)
+        if (task.numFailures >= lifecycleConfig.numFailuresAllowed - 1) {
+          log.warn("Too many errors (${lifecycleConfig.numFailuresAllowed}) " +
+            "fetching the task status. Giving up.")
+          publishExceptionEvent(task)
+          monitorRepository.delete(task)
+        } else {
+          monitorRepository.markFailureGettingStatus(task)
+        }
+      }
+  }
+
+  private fun orcaTaskIdToLink(task: MonitoredTask): String =
```

Comment:
Ha! Well. Good question. Both "link" fields are ambiguous. The thought is the link contains something that a monitor knows how to deal with. Constructing the spinnaker "link" for the user requires reading from the config to get the base URL. I thought that fit better in the handlers rather than the place we emit events.

It might be better to have this function in the place where we summarize all the events for the API, and that way we could just take anything that looks like an orca task id and make it into a user link. but, what if something looks like a uid but isn't for orca? the monitor for that thing would know about it, and would be able to construct the right link for the user. So that's why I decided to allow link to be any text and rely on the monitor to update it to a user friendly link.